solutions/cfgym-102694C.mdx

```diff
@@ -0,0 +1,240 @@
+---
+id: cf-831C
+source: CF
+title: Sloth Naptime
+author: Sofia Yang
+---
+
+[Announcement With Official Editorial](https://codeforces.com/blog/entry/81527)
+
+**Time Complexity:** $\mathcal{O}(N \log N)$
+
+The sloth will always try to move upwards from the start node $(st)$ to the
+least common ancestor of the start and end nodes $(lca)$ and then move downwards
+from the least common ancestor towards the end node $(end)$.
+
+Let $dist(i, j)$ be the distance from node i to node j.
+
+So if the sloth has energy amount $e$, there are 3 separate cases:
+
+1. $e \geq dist(st, end)$. In this case the result will be the sloth reaching the
+   end.
+
+2. $e < dist(st, end)$ and $e < dist(st, lca)$ in which the result will be the
+   sloth reaching the $e$th parent of $st$.
+
+3. $e < dist(st, end)$ and $e \geq dist(st, lca)$ in which the result will be the
+   sloth reaching the $(e - dist(st, lca))$th parent of $end$.
+
+So first, we can can run DFS once to find the depth of every node.
+
+Then we can create a matrix $anc$ to store the ancestors of each node, where
+$anc[i][j]$ is the $2^j$th ancestor of node $i$.
+
+After that, we can use binary lifting to answer each query by finding the least
+common ancestor of the start and end nodes, and then finding the right node that
+the sloth will reach.
+
+## Implementation
+
+<LanguageSection>
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+public class SlothNaptime {
+	public static final int MAXN = (int) 3e5;
```

Comment:
```suggestion
	public static final int MAXN = (int) 3e5 + 5;
```

Reply:
Ok Sans, so should it be 300000 or  300005? 
Since it's indexed by 0, 300000 it should work, but I guess I can change it back to 300005?

Reply:
3e5+5, since its better to add some buffer just in case

Reply:
nononono
if your code only works w/ a buffer it means something's wrong w/ your code

Reply:
hm? no thats not nec. true. Its just to prevent some random memory errors.

Reply:
wdym "random"? errors are 99.99999% not random

Reply:
@SansPapyrus683 it's just better to include a +5 or a +10.

Reply:
but like why lol

Reply:
Honestly, I don't care either way, but I can confirm that 3*10^5 works.
So there's no errors
![untitled](https://user-images.githubusercontent.com/48630514/139696646-d239b077-b419-42b0-b832-0d7ec8223f95.png)
.